modular_zubbers/modules/plexagon_selfserve/code/plexagon_selfserve.dm

```diff
@@ -0,0 +1,282 @@
+/datum/computer_file/program/crew_self_serve/proc/job_is_CMD_or_SEC()
+	if(!authenticated_card)
+		return FALSE
+
+	var/datum/id_trim/job/current_trim = authenticated_card.trim
+	var/datum/job/clocked_in_job = current_trim.job
+	if((/datum/job_department/command in clocked_in_job.departments_list) || (/datum/job_department/security in clocked_in_job.departments_list))
+		return TRUE
+
+	return FALSE
+
+/// Is the inserted ID on cooldown? returns TRUE if the ID has a cooldown
+/datum/computer_file/program/crew_self_serve/proc/id_cooldown_check()
+	if(!authenticated_card)
+		return FALSE
+
+	var/datum/component/off_duty_timer/id_component = authenticated_card.GetComponent(/datum/component/off_duty_timer)
+	if(!id_component)
+		return FALSE
+
+	if(!id_component.on_cooldown)
+		return FALSE
+
+	return TRUE
+
+/// Is the inserted ID on cooldown? returns TRUE if the ID has a cooldown
+/datum/computer_file/program/crew_self_serve/proc/id_locked_check()
+	if(!authenticated_card)
+		return FALSE
+
+	var/datum/component/off_duty_timer/id_component = authenticated_card.GetComponent(/datum/component/off_duty_timer)
+	if(!id_component)
+		return FALSE
+
+	if (id_component.on_cooldown)
+		return TRUE
+
+	if(id_component.hop_locked)
+		return TRUE
+
+	return FALSE
+
+/// Is the inserted ID off-duty? Returns true if the ID is off-duty
+/datum/computer_file/program/crew_self_serve/proc/off_duty_check()
+	if(!authenticated_card)
+		return FALSE
+
+	var/datum/component/off_duty_timer/id_component = authenticated_card.GetComponent(/datum/component/off_duty_timer)
+	if(!id_component)
+		return FALSE
+
+	return TRUE
+
+///Ejects the ID stored inside of the parent machine, if there is one.
+/datum/computer_file/program/crew_self_serve/proc/eject_inserted_id(mob/recepient)
+	if(!authenticated_card || !recepient)
+		return FALSE
+
+	authenticated_card.forceMove(computer.loc)
+	recepient.put_in_hands(authenticated_card)
+
+	authenticated_card = FALSE
+	computer.update_appearance()
+	computer.update_static_data_for_all_viewers()
+	playsound(computer, 'sound/machines/terminal_insert_disc.ogg', 50, FALSE)
+
+	return TRUE
+
+/datum/computer_file/program/crew_self_serve/kill_program(mob/user)
+	computer.crew_manifest_update = FALSE
+	if(!isnull(authenticated_card))
+		authenticated_card = null
+
+	return ..()
+
+/datum/computer_file/program/crew_self_serve/ui_act(action, params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+
+	authenticate(computer.computer_id_slot)
+	switch(action)
+		if("PRG_change_status")
+			if(off_duty_check())
+				if (!authenticated_card)
+					return;
+
+				if(!(clock_in()))
+					return
+				log_admin("[key_name(usr)] clocked in as \an [authenticated_card.assignment].")
+
+				var/datum/mind/user_mind = usr.mind
+				if(user_mind)
+					user_mind.clocked_out_of_job = FALSE
+
+			else
+				if (!authenticated_card)
+					return;
+
+				log_admin("[key_name(usr)] clocked out as \an [authenticated_card.assignment].")
```

Comment:
optimally any logging should be done on the central proc that both the program and stationary computer do, IE, on the jobs subsystem